### PR TITLE
Test extension options in DocGen

### PR DESCRIPTION
**Pull request by magicDGS**
- Add tests for non-html index/features extensions
- New templates for testing extensions: markdown and yml
- Fix bug: no recognized -index-output-extension option in HelpDoclet

### CODE REVIEW

**Comment by magicDGS** [comment]
Sorry @cmnbroad, I introduced a bug in #60 because I've overseen how to implement the new options. This add the tests promised there and fix the bug. Thank you very much!

**Comment by magicDGS** [comment]
Please, can you have a look to this one @cmnbroad? This was a bug that should be fixed before I can use the index-extension feature!

File: src/test/java/org/broadinstitute/barclay/help/DocumentationGenerationIntegrationTest.java [hunk 1]
```diff
@@ -42,38 +42,46 @@
...
+                {TestDoclet.class, new File("src/test/resources/org/broadinstitute/barclay/help/templates/TestDoclet"), new File("src/test/resources/org/broadinstitute/barclay/help/expected/TestDoclet"), "html", "html"},
+                // custom templates with different extension
+                {HelpDoclet.class, new File("src/test/resources/org/broadinstitute/barclay/help/templates/markdown/"), new File("src/test/resources/org/broadinstitute/barclay/help/expected/markdown"), "yml", "md"}
         };
     }
 
     @Test(dataProvider = "getDocGenTestParams")
-    public void testDocGenRoundTrip(final Class<?> docletClass, final File templatesFolder, final File expectedDir) throws IOException {
+    public void testDocGenRoundTrip(final Class<?> docletClass, final File templatesFolder, final File expectedDir,
+            final String indexFileExtension, final String outputFileExtension) throws IOException {
```
Comment (cmnbroad) [inline src/test/java/org/broadinstitute/barclay/help/DocumentationGenerationIntegrationTest.java]: It seem unnecessary to add an additional set of input templates and expected files just so we can test that the output extension works correctly. Can you just re-use the existing input templates and expected files, and just change the test to know what output extension to expect ? The contents should be the same no matter what the extension.

File: src/main/java/org/broadinstitute/barclay/help/HelpDoclet.java [hunk 2]
```diff
@@ -92,6 +92,7 @@
      * {@link #startProcessDocs(RootDoc)}.
      */
      public static boolean start(final RootDoc rootDoc) throws IOException {
+         System.out.println(Arrays.toString(rootDoc.options()));
```
Comment (magicDGS) [inline src/main/java/org/broadinstitute/barclay/help/HelpDoclet.java]: Debugging remains. Removed!

File: src/test/java/org/broadinstitute/barclay/help/DocumentationGenerationIntegrationTest.java (hunk 1 above)
Comment (magicDGS) [inline src/test/java/org/broadinstitute/barclay/help/DocumentationGenerationIntegrationTest.java]: True. I refactored the tests to make it simpler and re-use test files.

**Comment by magicDGS** [comment]
Back to you, @cmnbroad!

File: src/test/java/org/broadinstitute/barclay/help/DocumentationGenerationIntegrationTest.java [hunk 3]
```diff
@@ -30,7 +29,15 @@
             "-cp", System.getProperty("java.class.path")
     );
 
-    private static String[] docArgList(final Class<?> docletClass, final File templatesFolder, final File outputDir) {
+    private static final List<String> DOCUMENTED_PHP_NAMES_WITHOUT_EXTENSION = Arrays.asList(
+            "org_broadinstitute_barclay_help_TestArgumentContainer.",
+            "org_broadinstitute_barclay_help_TestArgumentContainer.",
+            "org_broadinstitute_barclay_help_TestExtraDocs.",
+            "org_broadinstitute_barclay_help_TestExtraDocs."
```
Comment (cmnbroad) [inline src/test/java/org/broadinstitute/barclay/help/DocumentationGenerationIntegrationTest.java]: I don't think we need two of each of these...

File: src/test/java/org/broadinstitute/barclay/help/DocumentationGenerationIntegrationTest.java [hunk 4]
```diff
@@ -42,46 +49,49 @@
...
 
     @DataProvider
     public Object[][] getDocGenTestParams() {
         return new Object[][] {
                 // default doclet and templates
-                {HelpDoclet.class, new File("src/main/resources/org/broadinstitute/barclay/helpTemplates/"), new File("src/test/resources/org/broadinstitute/barclay/help/expected/HelpDoclet")},
+                {HelpDoclet.class, new File("src/main/resources/org/broadinstitute/barclay/helpTemplates/"), new File("src/test/resources/org/broadinstitute/barclay/help/expected/HelpDoclet"), "html", "html"},
                 // custom doclet and templates
-                {TestDoclet.class, new File("src/test/resources/org/broadinstitute/barclay/help/templates/TestDoclet"), new File("src/test/resources/org/broadinstitute/barclay/help/expected/TestDoclet")}
+                {TestDoclet.class, new File("src/test/resources/org/broadinstitute/barclay/help/templates/TestDoclet"), new File("src/test/resources/org/broadinstitute/barclay/help/expected/TestDoclet"), "html", "html"}
```
Comment (cmnbroad) [inline src/test/java/org/broadinstitute/barclay/help/DocumentationGenerationIntegrationTest.java]: We're still only testing with html outputs. To have a valid test, we need to set the index output extension to something other than the default.

File: src/test/java/org/broadinstitute/barclay/help/DocumentationGenerationIntegrationTest.java [hunk 5]
```diff
@@ -42,46 +49,49 @@
...
         // creates a temp output directory
         final File outputDir = Files.createTempDirectory(docletClass.getName()).toAbsolutePath().toFile();
         outputDir.deleteOnExit();
 
         // run javadoc with the the custom doclet
-        com.sun.tools.javadoc.Main.execute(docArgList(docletClass, templatesFolder, outputDir));
+        com.sun.tools.javadoc.Main.execute(docArgList(docletClass, templatesFolder, outputDir, indexFileExtension, outputFileExtension));
+
+        // Compare index files
+        Assert.assertTrue(filesContentsIdentical(new File(expectedDir, "index." + indexFileExtension), new File(outputDir, "index.html")));
```
Comment (cmnbroad) [inline src/test/java/org/broadinstitute/barclay/help/DocumentationGenerationIntegrationTest.java]: Not your fault, but I just noticed that this reverses the order of expected/actual arguments from the way they're declared by filesContentsIdentical.

File: src/test/java/org/broadinstitute/barclay/help/DocumentationGenerationIntegrationTest.java (hunk 3 above)
Comment (magicDGS) [inline src/test/java/org/broadinstitute/barclay/help/DocumentationGenerationIntegrationTest.java]: True, copy-paste leftovers...

File: src/test/java/org/broadinstitute/barclay/help/DocumentationGenerationIntegrationTest.java (hunk 4 above)
Comment (magicDGS) [inline src/test/java/org/broadinstitute/barclay/help/DocumentationGenerationIntegrationTest.java]: I think that I haven't commit that change. Anyway, it is in your commit.

File: src/test/java/org/broadinstitute/barclay/help/DocumentationGenerationIntegrationTest.java (hunk 5 above)
Comment (magicDGS) [inline src/test/java/org/broadinstitute/barclay/help/DocumentationGenerationIntegrationTest.java]: Done in your commit, I guess.

**Comment by magicDGS** [comment]
Back to you @cmnbroad - I added your commit and did a very minor change (indexFileName and jsonFileExtension are know private static).

**Comment by cmnbroad** [comment]
Thanks @magicDGS!

**Comment by magicDGS** [comment]
Thanks to you for reviewing, @cmnbroad!

**Pull request merged**